### PR TITLE
Sol resolver fixes v3

### PR DESCRIPTION
**Pull request by LogvinovLeon**
## Description

This PR is a try to make the version locking work.

In particular:
* `NPMResolver` now has more rigorous checks. It checks that all the resolved dependencies are listed in package.json and the versions match the range required.
* `ReentrantERC20Token` was moved to the exchange package

## Types of changes

* Bug fix (non-breaking change which fixes an issue)

* New feature (non-breaking change which adds functionality)

* Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,6 +43,8 @@
         "generate_doc": "node ./packages/monorepo-scripts/lib/doc_generate_and_upload.js",
         "test:generate_docs:circleci": "for i in ${npm_package_config_packagesWithDocPages}; do yarn generate_doc --package $i --shouldUpload false --isStaging true || break -1; done;",
         "bundlewatch": "bundlewatch",
+        "comment:postinstall": "HACK: For some reason `yarn` is not setting up symlinks properly for some modules. We temporarily set them manually. Remove this after V3 refactor is complete.",
+        "postinstall": "rm -rf `pwd`/contracts/exchange-libs/node_modules/@0x; mkdir `pwd`/contracts/exchange-libs/node_modules/@0x; ln -s `pwd`/contracts/utils `pwd`/contracts/exchange-libs/node_modules/@0x/contracts-utils",
```
Comment (abandeali1) [inline package.json]: Man this is ugly. Will it always be these specific packages that are the issue? Is there really no other way to resolve this issue?

File: packages/sol-compiler/src/compiler.ts [hunk 2]
```diff
@@ -113,7 +113,7 @@ export class Compiler {
         this._nameResolver = new NameResolver(this._contractsDir);
         const resolver = new FallthroughResolver();
         resolver.appendResolver(new URLResolver());
-        resolver.appendResolver(new NPMResolver(this._contractsDir));
+        resolver.appendResolver(new NPMResolver(process.cwd()));
```
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/sol-compiler/src/compiler.ts]: Yes, but we also had that before. Sol-compiler should be run from within your project directory

File: package.json (hunk 1 above)
Comment (LogvinovLeon) [reply package.json]: It is ugly, but we haven't found any better way around it. We can remove that hack and try to debug yarn. We've done the same for v2.

File: packages/sol-compiler/src/compiler.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/sol-compiler/src/compiler.ts]: We do `path.resolve` for the `contractsDir` above

File: packages/sol-resolver/src/resolvers/npm_resolver.ts [hunk 3]
```diff
@@ -1,42 +1,86 @@
...
+        const packageJSONPath = path.join(packagePath, 'package.json');
+        const packageJSON = JSON.parse(fs.readFileSync(packageJSONPath).toString());
+        const packageVersion = packageJSON.version;
+        return packageVersion;
+    }
     constructor(packagePath: string) {
         super();
         this._packagePath = packagePath;
+        const packageJSONPath = path.join(this._packagePath, 'package.json');
+        const packageJSON = JSON.parse(fs.readFileSync(packageJSONPath).toString());
```
Comment (fabioberger) [inline packages/sol-resolver/src/resolvers/npm_resolver.ts]: Let's dedup this code with the above. Perhaps a helper method that returns the `package.json` for a path?

File: packages/sol-resolver/src/resolvers/npm_resolver.ts [hunk 4]
```diff
@@ -1,42 +1,86 @@
...
-                    : path.join(packageScopeIfExists, packageName);
-                const lookupPath = path.join(currentPath, 'node_modules', packagePath, pathWithinPackage);
-                if (fs.existsSync(lookupPath) && fs.lstatSync(lookupPath).isFile()) {
-                    const fileContent = fs.readFileSync(lookupPath).toString();
-                    return { source: fileContent, path: importPath, absolutePath: lookupPath };
-                }
-                currentPath = path.dirname(currentPath);
-            }
+        if (importPath.startsWith('/')) {
+            return undefined;
```
Comment (fabioberger) [inline packages/sol-resolver/src/resolvers/npm_resolver.ts]: Can you explain this conditional? Why is this special-cased?

File: packages/sol-resolver/src/resolvers/npm_resolver.ts [hunk 5]
```diff
@@ -1,42 +1,86 @@
...
+import * as semver from 'semver';
 
 import { ContractSource } from '../types';
 
 import { Resolver } from './resolver';
 
 export class NPMResolver extends Resolver {
     private readonly _packagePath: string;
+    private readonly _dependencies: { [dependencyName: string]: string };
+    private static _lookupPath(startingPath: string, lookupTarget: string): string | undefined {
```
Comment (fabioberger) [inline packages/sol-resolver/src/resolvers/npm_resolver.ts]:
```suggestion
    private static _lookupPathIfExists(startingPath: string, lookupTarget: string): string | undefined {
```

File: packages/sol-resolver/src/resolvers/npm_resolver.ts [hunk 6]
```diff
@@ -1,42 +1,86 @@
...
+        const dependencyLookupTarget = path.join('node_modules', dependencyName, internalPath);
+        const dependencyLookupResult = NPMResolver._lookupPath(this._packagePath, dependencyLookupTarget);
+        if (_.isUndefined(dependencyLookupResult)) {
+            return undefined;
+        }
+        if (_.isUndefined(this._dependencies[dependencyName])) {
+            throw new Error(
+                `Lookup for ${dependencyName} started from ${this._packagePath} failed. Not specified in package.json`,
+            );
+        }
```
Comment (fabioberger) [inline packages/sol-resolver/src/resolvers/npm_resolver.ts]: Move this check up after the call to `NPMResolver._parseDependencyImport`. It does not depend on looking up the path and we should fail as fast as possible.

**Pull request merged**